### PR TITLE
PIM-10887 :  Prevent channel creation on validation error during import

### PR DESCRIPTION
**Pull request by Beninho**
**Description (for Contributor and Core Developer)**

Prevent channel creation on validation error during import

- Create channel processor from generic to remove association between channels and locales on new channels
- Update `ChannelLocaleSubscriber` in order to save locale updates only on **POST** events 
- Remove cascade persist between from Locale To channels
- Update acceptance test to check activation/deactivation of locale

**Definition Of Done (for Core Developer only)**

- [X] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [X] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: tests/back/Pim/Enrichment/EndToEnd/RemoveLocaleFromChannelEndToEnd.php [hunk 1]
```diff
@@ -79,6 +70,54 @@ public function test_that_removing_a_locale_from_a_channel_recomputes_the_produc
...
+                'code' => 'my_channel',
+            ])
+        );
+        $response = $this->client->getResponse();
+        $this->assertEquals(Response::HTTP_NO_CONTENT, $response->getStatusCode());
+        $this->waitForJobExecutionToEnd();
+
+        $this->assertEquals($this->isActivatedLocale('br_FR'), false, 'locale must be deactivated');
+    }
+
```
Comment (Beninho) [inline tests/back/Pim/Enrichment/EndToEnd/RemoveLocaleFromChannelEndToEnd.php:120]: test with invalid channel code

File: src/Akeneo/Channel/back/Infrastructure/EventListener/ChannelLocaleSubscriber.php [hunk 2]
```diff
@@ -58,89 +41,59 @@ public function __construct(
...
 
-        if (!empty($updatedLocales)) {
-            $this->saver->saveAll($updatedLocales);
+        $oldLocales = $this->repository->getDeletedLocalesForChannel($channel);
+        foreach ($oldLocales as $locale) {
+            if ($locale->hasChannel($channel)) {
+                $locale->removeChannel($channel);
+            }
+            $this->updatedLocales[$channel->getCode()][$locale->getCode()] = $locale;
+            $this->localesRemovedFromChannel[$channel->getCode()][$locale->getCode()] = $locale->getCode();
```
Comment (pierallard) [inline src/Akeneo/Channel/back/Infrastructure/EventListener/ChannelLocaleSubscriber.php:69]: What's the point to have this structure and not

```suggestion
            $this->localesRemovedFromChannel[$channel->getCode()][] = $locale->getCode();
```

File: src/Akeneo/Channel/back/tests/Specification/Infrastructure/Processor/Denormalization/ChannelProcessorSpec.php [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+ */
+class ChannelProcessorSpec extends ObjectBehavior
+{
+    public function let(
+        IdentifiableObjectRepositoryInterface $repository,
+        SimpleFactory $factory,
+        ObjectUpdaterInterface $updater,
+        ValidatorInterface $validator,
+        ObjectDetacherInterface $objectDetacher
+
```
Comment (pierallard) [inline src/Akeneo/Channel/back/tests/Specification/Infrastructure/Processor/Denormalization/ChannelProcessorSpec.php:39]:
```suggestion
```

File: src/Akeneo/Channel/back/tests/Specification/Infrastructure/Processor/Denormalization/ChannelProcessorSpec.php [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    {
+        $this->shouldImplement(ItemProcessorInterface::class);
+        $this->shouldImplement(StepExecutionAwareInterface::class);
+    }
+
+    public function it_updates_an_existing_channel(
+        $repository,
+        $updater,
+        $validator,
+        ChannelInterface $channel,
```
Comment (pierallard) [inline src/Akeneo/Channel/back/tests/Specification/Infrastructure/Processor/Denormalization/ChannelProcessorSpec.php:59]: Please add the types of these variables (and in the next functions too)

File: src/Akeneo/Channel/back/tests/Specification/Infrastructure/Processor/Denormalization/ChannelProcessorSpec.php [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+        $this
+            ->shouldThrow(InvalidItemFromViolationsException::class)
+            ->during(
+                'process',
+                [$values]
+            );
+    }
+
+    protected function getValues(): array
```
Comment (pierallard) [inline src/Akeneo/Channel/back/tests/Specification/Infrastructure/Processor/Denormalization/ChannelProcessorSpec.php:253]: why protected and not private ?

File: tests/back/Pim/Enrichment/EndToEnd/RemoveLocaleFromChannelEndToEnd.php [hunk 6]
```diff
@@ -202,4 +275,10 @@ private function createFamilyWithRequirement(
         $this->validate($family);
         $this->get('pim_catalog.saver.family')->save($family);
     }
+
+    private function isActivatedLocale(string $localeCode): bool
+    {
+        $activatedLocales = $this->localeRepository->getActivatedLocaleCodes();
+        return \in_array($localeCode, $activatedLocales);
```
Comment (pierallard) [inline tests/back/Pim/Enrichment/EndToEnd/RemoveLocaleFromChannelEndToEnd.php:282]:
```suggestion
        $activatedLocales = $this->localeRepository->getActivatedLocaleCodes();
        
        return \in_array($localeCode, $activatedLocales);
```

**Review by pierallard — APPROVED** [review]

File: src/Akeneo/Channel/back/Infrastructure/EventListener/ChannelLocaleSubscriber.php (hunk 2 above)
Comment (Beninho) [reply src/Akeneo/Channel/back/Infrastructure/EventListener/ChannelLocaleSubscriber.php:69]: It's in order not to have several times the same locale

**Review by antoinearrive — APPROVED** [review]

**Pull request merged**